File: java/21_language_features.java
// ============================================================
// Topic:  Language Features  (FOCUS: Streams, var, text blocks, switch, record, Optional)
// Summary: Modern Java syntax: Stream API, var, text blocks, switch expressions,
//          records (16+), Optional.
// Run:     java java/21_language_features.java   (Java 11+ single-file launch;
//          records require Java 16+)
//
// KEY DIFFERENCES:
//   Python:  list comprehensions / map+filter; no var; triple-quoted strings;
//            match/case (3.10+); dataclasses; Optional typing only
//   TS/JS:   .map/.filter on arrays; const/let inference; template literals;
//            switch; interfaces/types; null/undefined (no Optional type)
//   Rust:    iterators .map/.filter; let inference; raw strings r#"..."#;
//            match; struct + derive; Option<T>
//   Java:    Stream API; var (10+); text blocks (15+); switch -> (14+);
//            record (16+); java.util.Optional
//   C++:     ranges/views (C++20); auto; raw string literals; switch; struct;
//            std::optional
// ============================================================

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class LanguageFeatures {

    // ----- record (Java 16+) — compact immutable data carrier -----
    record Person(String name, int age) {}

    public static void main(String[] args) {
        // ----- 1. Streams map/filter -----
        String evens = List.of(1, 2, 3, 4, 5, 6).stream()
                .filter(n -> n % 2 == 0)
                .map(String::valueOf)
                .collect(Collectors.joining(" "));
        System.out.println("stream: " + evens);                  // Output: stream: 2 4 6

        // ----- 2. var local type inference (Java 10+) -----
        var greeting = "hello";
        System.out.println("var: " + greeting);                  // Output: var: hello

        // ----- 3. Text block (Java 15+) -----
        String block = """
                multi
                more
                """;
        String firstLine = block.lines().findFirst().orElse("").strip();
        System.out.println("block: " + firstLine);               // Output: block: multi

        // ----- 4. Switch expression with -> (Java 14+) -----
        int value = 5;
        String sign = switch (Integer.signum(value)) {
            case -1 -> "negative";
            case 0  -> "zero";
            default -> "positive";
        };
        System.out.println("switch: " + sign);                   // Output: switch: positive

        // ----- 5. record (Java 16+) -----
        Person bob = new Person("Bob", 30);
        System.out.println("record: " + bob.name() + " age=" + bob.age());
                                                                   // Output: record: Bob age=30

        // ----- 6. Optional — explicit absent/present wrapper -----
        Optional<Integer> maybe = Optional.of(7);
        System.out.println("optional: " + maybe.orElse(0));        // Output: optional: 7
    }
}

File: csharp/21_language_features.cs
// ============================================================
// Topic:  Language Features
// Summary: LINQ, auto-properties, pattern matching, interpolation,
//          nullable reference types, records with expressions.
// Run:     dotnet script csharp/21_language_features.cs
//
// KEY DIFFERENCES:
//   C#:     LINQ, records, pattern matching, nullable ref types
//   Python: list comprehensions, dataclasses, match/case (3.10+)
//   TS/JS:  Array.map/filter, optional chaining ?., template literals
//   Rust:   iterators, struct + derive, match, Option/?
//   Java:   Streams, records (16+), switch expressions, Optional
// ============================================================

#nullable enable

using System;
using System.Linq;

// ----- 1. LINQ — query syntax / method chain on IEnumerable -----
var nums = new[] { 1, 2, 3 };
var doubled = nums.Select(n => n * 2);                 // Select    | map each element
Console.WriteLine($"linq: {string.Join(" ", doubled)}"); // linq: 2 4 6

// ----- 2. Auto-properties — compiler-generated backing field -----
class Person
{
    public string Name { get; set; } = "Vetri";      // auto-prop | get/set without field
}
var person = new Person();
Console.WriteLine($"prop: {person.Name}");           // prop: Vetri

// ----- 3. Pattern matching switch — type and relational patterns -----
object value = 42;
string label = value switch
{
    int n when n > 10 => "big",                      // when      | guard on pattern
    int n => "small",
    _ => "other"
};
Console.WriteLine($"match: {label}");              // match: big

// ----- 4. String interpolation — $ prefix embeds expressions -----
string name = "Vetri";
Console.WriteLine($"fmt: hi {name}");                // fmt: hi Vetri

// ----- 5. Nullable reference types + null-conditional ?. -----
string? text = null;                                 // string?   | nullable annotation
int length = text?.Length ?? 0;                      // ?. ??     | null-safe access, default 0
Console.WriteLine($"nullsafe: {length}");            // nullsafe: 0

// ----- 6. record + with expression — immutable copy with changes -----
record Player(string Name, int Score);               // record    | positional, value equality
var original = new Player("Bob", 50);
var updated = original with { Name = "Alice", Score = 100 }; // with | non-destructive mutation
Console.WriteLine($"record: {updated.Name} score={updated.Score}"); // record: Alice score=100

// Cross-language note:
//   LINQ ≈ Python comprehensions / Rust iterators / Java Streams — lazy until enumerated.
//   records ≈ Python @dataclass / TS readonly tuples — structural equality by default.
//   with expressions ≈ spread copy ({ ...obj, name: "Alice" }) without mutating the original.
//   Nullable reference types are compile-time hints — runtime null still possible without checks.
//   Pattern matching switch subsumes classic switch; order matters when patterns overlap.